### PR TITLE
Bugfix #4336: ingester leaving pool

### LINKED ISSUE #4336: Distributed Ingest  bug
When running 
```
- chico(rate=10mb)*4 
 ```

Indexer eventually stops with as indexer-0 WAL hits its disk limit.
(Example at 8:42 UTC december 28.)

The WAL memory limit truncation on the other hand is ok.
After stopping pushing new documents, the faulty server does not go back to normal (the WAL stays at the limit).

This suggests a single shard is not being truncated ever.

Downloading the WAL informs us that the 
shard 4 of simian_chico_1072342904323664203 still contains documents.
3838250..3956732

The publish position in the metastore is 3838249. (which matches the WAL)

This position is alspo published correctly on chitchat. The logs tell us the last publish happended on indexer1 at 2023-12-28 08:44:49.105.

At 08:43:04.356, indexer-2 receives the assign shard command for shard 4.

https://console.cloud.google.com/logs/query;query=resource.labels.container_name%3D%22quickwit%22%20resource.labels.namespace_name%3D%22simian%22%20%0Alabels.%22k8s-pod%2Fstatefulset_kubernetes_io%2Fpod-name%22%3D%22simian-indexer-2%22;summaryFields=labels%252F%2522k8s-pod%252Fstatefulset_kubernetes_io%252Fpod-name%2522:false:32:beginning;cursorTimestamp=2023-12-28T08:44:14.307609241Z;startTime=2023-12-28T08:43:04.000Z;endTime=2023-12-28T08:44:57.384Z?_ga=2.235303489.-1376293965.1696294435&_gac=1.85231211.1703487344.CjwKCAiAyp-sBhBSEiwAWWzTnjRcqsvgsNhTwM-zzGj6rqUmDwRdBHD-yaauJLT3F1KcA3iiWXpPgxoCwZ8QAvD_BwE&hl=fr&authuser=1&project=quickwit-prod

We see `invalid publish token` errors.

---

What this is not:
- a truncation message not being sent.

What this is:
- a bug somewhere in the ingester or the ingesting pipeline that puts it in a state where indexing never progress on a shard.

Everything looks like somehow an error occurs, and indexing stops. 

023-12-28 08:43:28.171 
ingester 0 not available closing shard 4.
```
INFO 2023-12-28T08:43:28.171370938Z [labels."k8s-pod.statefulsetKubernetesIo.pod-name": simian-indexer-2] [resource.labels.containerName: quickwit] 2023-12-28T08:43:28.171Z ERROR quickwit_ingest::ingest_v2::fetch: ingester `simian-indexer-0` is not available: closing fetch stream client_id=indexer/simian-indexer-2/simian_chico_1072342904323664203:01HJQP9T4YVTSG528AH2R1JB0M/_ingest-source/01HJQP9T8GDXQ2YWT8Q8FGK8TR index_uid=simian_chico_1072342904323664203:01HJQP9T4YVTSG528AH2R1JB0M source_id=_ingest-source shard_id=4
  {
    "textPayload": "2023-12-28T08:43:28.171Z ERROR quickwit_ingest::ingest_v2::fetch: ingester `simian-indexer-0` is not available: closing fetch stream client_id=indexer/simian-indexer-2/simian_chico_1072342904323664203:01HJQP9T4YVTSG528AH2R1JB0M/_ingest-source/01HJQP9T8GDXQ2YWT8Q8FGK8TR index_uid=simian_chico_1072342904323664203:01HJQP9T4YVTSG528AH2R1JB0M source_id=_ingest-source shard_id=4",
    "insertId": "hm3q6a38ab365heg",
    "resource": {
      "type": "k8s_container",
      "labels": {
        "location": "us-east1-d",
        "cluster_name": "quickwit-prod-gke",
        "container_name": "quickwit",
        "project_id": "quickwit-prod",
        "namespace_name": "simian",
        "pod_name": "simian-indexer-2"
      }
    },
    "timestamp": "2023-12-28T08:43:28.171370938Z",
    "severity": "INFO",
    "labels": {
      "k8s-pod/app_kubernetes_io/component": "indexer",
      "k8s-pod/app_kubernetes_io/name": "quickwit",
      "k8s-pod/controller-revision-hash": "simian-indexer-5658df6d98",
      "compute.googleapis.com/resource_name": "gke-quickwit-prod-gk-quickwit-prod-gk-c4ded138-updk",
      "k8s-pod/statefulset_kubernetes_io/pod-name": "simian-indexer-2",
      "k8s-pod/app_kubernetes_io/instance": "release-name"
    },
    "logName": "projects/quickwit-prod/logs/stdout",
    "receiveTimestamp": "2023-12-28T08:43:32.440340296Z"
  }

```

### PR DESCRIPTION
**Pull request by fulmicoton**
Avoid evicting ingester node from IngesterPool when facing or a transport error.

    Before when facing such error case, we were removing the faulty nodes from the pool, and
    Nn code path would re-add it to the ingester pool. The ingester pool is also used by the ingest
    source resulting in bug #4336.

    After this patch:

    When facing a transport error, we assume the targetted node is
    unreachable and chitchat has just not detected this just yet.

    In an ideal world we would inform chitchat about this, but it is a bit
    difficult to do codewise.

    Instead, we register the leader as unavailable for the span of the
    workbench. It will then react as if it was out of the pool for
    subsequent retries.

    A GetOrCreatedShard will carry the information that the node was
    unavailable, and the control plane will attempt to create a shard on a
    different node

    A timeout on the other hand is treated as a normal retryable error.

    Closes #4336

### CODE REVIEW

File: quickwit/quickwit-ingest/src/ingest_v2/router.rs [hunk 1]
```diff
@@ -267,9 +260,10 @@ impl IngestRouter {
                             for subrequest_id in persist_summary.subrequest_ids {
                                 workbench.record_no_shards_available(subrequest_id);
                             }
-                            self.ingester_pool.remove(&persist_summary.leader_id);
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/router.rs:270]: this is the bugfix.

File: quickwit/quickwit-ingest/src/ingest_v2/ingester.rs [hunk 2]
```diff
@@ -273,9 +273,7 @@ impl Ingester {
                     "failed to create mrecordlog queue `{}`: {}",
                     queue_id, io_error
                 );
-                return Err(IngestV2Error::IngesterUnavailable {
-                    ingester_id: shard.leader_id.into(),
-                });
+                return Err(IngestV2Error::Internal(format!("Io Error: {io_error}")));
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/ingester.rs:276]: mrecordlog io error  are treated as `InternalError`

File: quickwit/quickwit-ingest/src/ingest_v2/router.rs [hunk 3]
```diff
@@ -122,25 +119,31 @@ impl IngestRouter {
...
         // `closed_shards` and `unavailable_leaders` are populated by calls to `has_open_shards`
         // as we're looking for open shards to route the subrequests to.
         let mut closed_shards: Vec<ShardIds> = Vec::new();
-        let mut unavailable_leaders: HashSet<NodeId> = HashSet::new();
+        let unavailable_leaders: &mut HashSet<NodeId> = &mut workbench.unavailable_leaders;
 
         let state_guard = self.state.read().await;
 
-        for subrequest in subrequests {
+        for subrequest in workbench.subworkbenches.values().filter_map(|subworbench| {
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/router.rs:134]: that's lame. I inlined this to workaround the mutability.

File: quickwit/quickwit-ingest/src/ingest_v2/workbench.rs [hunk 4]
```diff
@@ -27,24 +27,31 @@ use quickwit_proto::ingest::router::{
...
 pub(super) struct IngestWorkbench {
-    pub subworkbenches: HashMap<SubrequestId, IngestSubworkbench>,
+    pub subworkbenches: BTreeMap<SubrequestId, IngestSubworkbench>,
     pub num_successes: usize,
     /// The number of batch persist attempts. This is not sum of the number of attempts for each
     /// subrequest.
     pub num_attempts: usize,
     pub max_num_attempts: usize,
+    // List of leaders that have been marked as temporarily unavailable.
+    // This leaders have encounterred a transport error during an attempt and will be treated as if
```
Comment (guilload) [inline quickwit/quickwit-ingest/src/ingest_v2/workbench.rs:44]:
```suggestion
    // These leaders have encountered a transport error during an attempt and will be treated as if
```

File: quickwit/quickwit-ingest/src/ingest_v2/workbench.rs [hunk 5]
```diff
@@ -137,44 +134,39 @@ impl IngestWorkbench {
...
-        subworkbench.num_attempts += 1;
-        subworkbench.last_failure_opt = Some(SubworkbenchFailure::NoShardsAvailable);
+        self.record_failure(subrequest_id, SubworkbenchFailure::NoShardsAvailable);
+    }
+
+    /// Marks a node as unavailable for the span of the workbench.
+    ///
+    /// Remaining attempts will treat the node as if it was not in the ingester pool.
+    pub fn record_connection_error(&mut self, subrequest_id: SubrequestId) {
+        self.record_failure(subrequest_id, SubworkbenchFailure::ConnectionError);
```
Comment (guilload) [inline quickwit/quickwit-ingest/src/ingest_v2/workbench.rs:154]: Is a connection error semantically different than a transport error, and do we need to distinguish between them? If not, can we use the same name?

**Review by guilload — APPROVED** [review]

**Pull request merged**